pootle/apps/virtualfolder/models.py

```diff
@@ -347,19 +275,156 @@ def get_adjusted_pootle_path(self, pootle_path):
         trail = pootle_path.replace(lead, '').lstrip('/')
         return '/'.join([lead, self.name, trail])
 
-    def get_translate_url(self, pootle_path, **kwargs):
-        """Get the translate URL for this virtual folder in the given path."""
-        adjusted_path = self.get_adjusted_pootle_path(pootle_path)
-        lang, proj, dp, fn = split_pootle_path(adjusted_path)
 
+class VFolderCachedTreeItemManager(models.Manager):
+    use_for_related_fields = True
+
+    def get_queryset(self):
+        return super(VFolderCachedTreeItemManager, self) \
+            .get_queryset().select_related('vfolder')
+
+
+class VFolderCachedTreeItem(models.Model, CachedTreeItem):
```

Comment:
We can rename it to *VirtualFolderTreeItem* to keep it descriptive while dropping the *Cached* part. But it can still be confusing.

On the need for a separate model:

Lets suppose we add a `VirtualFolder` with location `/af/firefox/browser/` with a filter `chrome/browser/overview/one.po, chrome/browser/migrations/two.po`. But we want to have specific stats for the virtual folder in each path between the vfolder files and the root location. Specifically the vfolder will have different stats for path `/af/firefox/browser/chrome/browser/overview/` and path `/af/firefox/browser/chrome/browser/migrations/` and also for their parent path `/af/firefox/browser/chrome/browser/`.

So since the TreeItem class provides the stats caching infrastructure provided that a tree structure is used it felt more natural to have a separate class to keep those treeitems for the vfolders.

Having `VFolderCachedTreeItem` also allows to easily retrieve all the vfolders for a given `Directory` without need to filter the `VirtualFolder` instances based on if they have units below that directory. See https://github.com/translate/pootle/pull/3816/files#diff-2595359c0c549cb2373dc7ec23dc8cb2L133 compared to https://github.com/translate/pootle/pull/3816/files#diff-2595359c0c549cb2373dc7ec23dc8cb2R289

I must say I don't see how `VirtualFolder` and `VFolderCachedTreeItem` can be merged together without either:

- Going back to having methods that convert provided directory pootle_path and vfolder to the equivalent `VFolderCachedTreeItem` pootle_path or
- Replicating the `VirtualFolder` specific data across all the `VFolderCachedTreeItem` instances for each vfolder.

I am not sure if this clarifies this to you. Please ask again for further explanations if necessary.